### ISSUE TITLE
[Bug]: Flatpaks cannot see fonts installed in $HOME by the KDE Font Management KCM

### ISSUE THREAD

**Issue by sfalken** [flatpak/flatpak#5675]
### Checklist

- [X] I agree to follow the [Code of Conduct](https://github.com/flatpak/flatpak/blob/main/CODE_OF_CONDUCT.md) that this project adheres to.
- [X] I have searched the [issue tracker](https://www.github.com/flatpak/flatpak/issues) for a bug that matches the one I want to file, without success.
- [X] If this is an issue with a particular app, I have tried filing it in the appropriate issue tracker for the app (e.g. under https://github.com/flathub/) and determined that it is an issue with Flatpak itself.
- [X] This issue is not a report of a security vulnerability (see [here](https://github.com/flatpak/flatpak/blob/main/SECURITY.md) if you need to report a security issue).

### Flatpak version

1.15.6

### What Linux distribution are you using?

openSUSE

### Linux distribution version

Tumbleweed

### What architecture are you using?

x86_64

### How to reproduce

1. Install fonts as a user, running the Plasma Desktop, using the KDE Font Management "Control Panel"
2. Start a flatpak
3. Go to font selection in the flatpak application
4. Only "system" fonts will be available

### Expected Behavior

To have fonts installed as "user fonts" in `~/.fonts` through the Plasma Desktop "control panel" available to my flatpak applications for use.

### Actual Behavior

"User" fonts, installed with the KDE Plasma "Font Management" KCM are not available to flatpak'd applications

The plasma desktop creates both `~/.fonts` and `~/.local/share/fonts`  but when installing "user" fonts through the provided GUI tool, they are installed to `~/.fonts`

Flatpak attempts to mount `~/.local/share/fonts` first, and if it doesn't exist, then it moves on to `~/.fonts`

Since the plasma desktop (or it's control panel) is providing both directories, when flatpak sees the empty `~/.local/share/fonts` it mounts that, and doesn't ever consider `~/.fonts`   Per https://github.com/flatpak/flatpak/blob/d5f891e0035e50b24211688a7fa5f61924a2e51c/common/flatpak-run.c#L938-L963

### Additional Information

Obviously, the user can just symlink `~/.fonts` to `~/.local/share/fonts` and get around this, but it seems like a relatively small issue to fix.    Alternatively, KDE upstream could start installing to `~/.local/share/fonts` but I don't have a good insight into that.

**Comment by sfalken** [comment]
Just to follow up, if I do `mv ~/.local/share/fonts ~/.local/share/fonts.old` and restart the flatpak (org.kde.yakuake in this case), it mounts `~/.fonts` to `/run/host/user-fonts` just fine, and acts as expected.

**Comment by chrisawi** [comment]
It looks like it'd be pretty easy to mount both directories, but `~/.fonts` has been deprecated for a long time. I'd much rather see KDE fix their font installer to use the correct path.

I see there's already a bug filed: https://bugs.kde.org/show_bug.cgi?id=442976

If the powers that be think it's worth adjusting this behavior in flatpak, I can make a PR, but I'm not going to bother otherwise.

**Comment by Erick555** [comment]
If flatpak will mount new path then fontconfig in all runtimes need to be patched for that.

**Comment by smcv** [comment]
> It looks like it'd be pretty easy to mount both directories

It would be easy, but would not solve the problem, because programs inside the sandbox are only configured to read fonts from `/run/host/user-fonts` (and some other locations which are not immediately relevant here). This is what @Erick555 is referring to.

If you have both `~/.fonts` and `~/.local/share/fonts` and they are different, Flatpak can only choose one of them to mount onto `/run/host/user-fonts` - it can't put both of them in the same place.

It seems that `~/.fonts` was deprecated 11 years ago, while `~/.local/share/fonts` is non-deprecated, so I think Flatpak is correct to prefer `~/.local/share/fonts` if they both exist. Preferring `~/.fonts` would be a regression for users who have fonts installed in `~/.local/share/fonts`.

Similarly, we can't repurpose the other locations in `/run/host/` that the runtimes know about, because each of them already has a use, and changing their meaning would be a regression for users who are relying on having fonts installed in `/usr/share/fonts` or `/usr/local/share/fonts`.

I think the right solution would be for KDE to install fonts into the non-deprecated location `~/.local/share/fonts`.

Until that's fixed, a workaround is to make `~/.fonts` a symlink to `~/.local/share/fonts` so that it doesn't matter which one is preferred.

**Comment by sfalken** [comment]
Fair enough, I'll pursue a change with, or look at sending a PR to the upstream to correct this.

**Comment by sfalken** [comment]
https://bugs.kde.org/show_bug.cgi?id=442976

**Comment by chrisawi** [comment]
> It would be easy, but would not solve the problem, because programs inside the sandbox are only configured to read fonts from /run/host/user-fonts (and some other locations which are not immediately relevant here).

I had originally assumed that was the case, but then I figured the generated fontconfig xml would take care of it. The documentation is unclear, but from the surface-level implementation, `<remap-dir>` seems like it should be sufficient to add a directory, rather than needing to also use `<dir>`.

I do see that `/run/host/local-fonts` supposedly didn't work until it was added to the runtime, but I've just patched flatpak locally to mount `~/.fonts` at `/run/host/user-fonts2` and everything seems to work. The font I added appears in `fc-list` under the correct path, and LibreOffice can still see it after removing its filesystem permissions.

**Comment by Erick555** [comment]
Do you mean after creating `/run/host/user-fonts2` fontconfig used it without being listed in https://gitlab.com/freedesktop-sdk/freedesktop-sdk/-/blob/master/files/fontconfig/50-flatpak.conf#L15 ? How?

**Comment by chrisawi** [comment]
flatpak wrote `/run/host/font-dirs.xml` containing:

    <remap-dir as-path="/var/home/chrisw/.fonts">/run/host/user-fonts2</remap-dir>

which is `<include>`d in that file from the runtime.

```
[📦 org.freedesktop.Platform ~]$ fc-list | grep Sixty
/run/host/user-fonts2/Sixtyfour-Regular.ttf: Sixtyfour:style=Regular
```

**Comment by smcv** [comment]
Hmm, perhaps `/run/host/font-dirs.xml` obsoletes the hard-coded handling of `/run/host/user-fonts` in the runtimes?

If it does, I'd be happy to review a merge request to add a mapping for `~/.fonts` in addition to `~/.local/share/fonts` if they both exist but are distinct directories, perhaps as `/run/host/user-dotfonts`.

The way to tell that they're distinct directories would be to `stat()` them both (which can replace the `g_file_query_exists()` call, too), and check whether the `st_dev` and/or `st_ino` is different, similar to how `portal/flatpak-portal.c` does an equivalent check.

**Comment by Erick555** [comment]
Hm font-dirs.xml support in [flatpak](https://github.com/flatpak/flatpak/pull/2635) is pretty ancient but [runtime](https://gitlab.com/freedesktop-sdk/freedesktop-sdk/-/commit/1de9095053f704871886deb6c433635aafc98062) support was added much later. It's still available in all supported runtimes.

**Comment by smcv** [comment]
> Hm font-dirs.xml support in [flatpak](https://github.com/flatpak/flatpak/pull/2635) is pretty ancient but [runtime](https://gitlab.com/freedesktop-sdk/freedesktop-sdk/-/commit/1de9095053f704871886deb6c433635aafc98062) support was added much later. It's still available in all supported runtimes.

OK, so to keep supporting older runtimes, we still need to be using the paths that those older runtimes hard-code, like `/run/host/user-fonts`, for which what we said above is true: if the user has both `~/.local/share/fonts` or `~/.fonts` and they are distinct, then we have to choose *one* of those two paths to be `/run/host/user-fonts` (and I think the right one to choose is `~/.local/share/fonts`).

But, we could potentially have an extra path where `~/.fonts` gets mounted *as well*, if someone wants to write the code for that. This would only work in newer runtimes.

**Comment by tomachinz** [comment]
> ~/.local/share/fonts

I just renamed mine to fonts_TOBEDELETED and now a whole range of flatpaks have started working again re: fonts. Then I hit "WhatsApp for Linux" which still looks like this:
![image](https://github.com/flatpak/flatpak/assets/1026291/1552b8b4-fa54-4e80-adaa-3779414ea7bc)

If the .local is the more common way maybe I will reverse my symlink potentially, but am about to add symlink pointing to ~/.fonts to learn more about this flatpakery behaviour.

**Comment by bugQ** [comment]
> https://bugs.kde.org/show_bug.cgi?id=442976

Looks like this was fixed in Plasma 6.2!
